include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -1276,319 +1268,377 @@ void solveRank1ProblemDist(CommSender&& row_comm, CommSender&& col_comm, const S
-            const SizeType j_el = dist_sub.global_element_from_local_element<Coord::Col>(j_el_lc);
-            const SizeType js_el = i6[j_el];
-
-            for (SizeType i_el_lc = 0; i_el_lc < m_el_lc; ++i_el_lc) {
-              const SizeType i_el = dist_sub.global_element_from_local_element<Coord::Row>(i_el_lc);
-              const SizeType is_el = i4[i_el];
-
-              // skip if deflated
-              if (is_el >= k)
-                continue;
-
-              // skip if originally it was on the diagonal
-              if (is_el == js_el)
-                continue;
-
-              const SizeType linear_lc = dist_extra::local_tile_linear_index(
-                  dist_sub, {dist_sub.local_tile_from_local_element<Coord::Row>(i_el_lc),
-                             dist_sub.local_tile_from_local_element<Coord::Col>(j_el_lc)});
-              const TileElementIndex ij_tl = dist_sub.tile_element_index({i_el, j_el});
-
-              w[i_el_lc] *=
-                  q[to_sizet(linear_lc)](ij_tl) / (d_ptr[to_sizet(is_el)] - d_ptr[to_sizet(js_el)]);
-            }
-          }
-        }
-
-        barrier_ptr->arrive_and_wait(barrier_busy_wait);
-
-        // STEP 2c: reduce, then finalize computation with sign and square root (single-thread)
-        if (thread_idx == 0) {
-          // local reduction from all bulk workers
-          for (SizeType i_el_lc = 0; i_el_lc < m_el_lc; ++i_el_lc) {
-            for (std::size_t tidx = 1; tidx < nthreads; ++tidx) {
-              const T* w_partial = ws_cols[tidx]();
-              w[i_el_lc] *= w_partial[i_el_lc];
-            }
-          }
-
-          tt::sync_wait(ex::when_all(row_comm_chain(),
-                                     ex::just(MPI_PROD, common::make_data(w, m_el_lc))) |
-                        transformMPI(all_reduce_in_place));
+      ex::transfer(hp_scheduler) |
+      ex::let_value([n, dist_sub, bcast_evals, all_reduce_in_place, hp_scheduler](
+                        auto& row_comm_wrapper, auto& col_comm_wrapper, const SizeType k,
+                        const SizeType k_lc, const auto& rho, const auto& d_tiles, auto& z_tiles,
+                        const auto& eval_tiles, const auto& i4_tiles_arr, const auto& i6_tiles_arr,
+                        const auto& i2_tiles_arr, const auto& evec_tiles, auto& ws_cols, auto& ws_row) {
+        using pika::execution::thread_priority;
+
+        const std::size_t nthreads = [dist_sub, k_lc] {
+          const std::size_t workload = to_sizet(dist_sub.localSize().rows() * k_lc);
+          const std::size_t workload_unit = 2 * to_sizet(dist_sub.tile_size().linear_size());
+
+          const std::size_t min_workers = 1;
+          const std::size_t available_workers = getTridiagRank1NWorkers();
+
+          const std::size_t ideal_workers = util::ceilDiv(to_sizet(workload), workload_unit);
+          return std::clamp(ideal_workers, min_workers, available_workers);
+        }();
```

Comment:
Just pointing out the logic, which might deserve a comment/note.

Number of "ideal" threads is calculated dynamically as a function of how much work rank1-solver has to do. The work is measured in terms of local problem size and a thread should have at least 2 tiles to work on.

This ideal number of threads is then used to determine how many actual threads can be used, by checking the range defined by configuration parameters (at least 1, at most config parameter).